File: libsrc/spatialdata/spatialdb/AnalyticDB.cc
// =================================================================================================
// This code is part of SpatialData, developed through the Computational Infrastructure
// for Geodynamics (https://github.com/geodynamics/spatialdata).
//
// Copyright (c) 2010-2025, University of California, Davis and the SpatialData Development Team.
// All rights reserved.
//
// See https://mit-license.org/ and LICENSE.md and for license information.
// =================================================================================================

#include <portinfo>

#include "AnalyticDB.hh" // Implementation of class methods

// Include ios here to avoid some Python/gcc issues
#include <ios>

#include "spatialdata/units/Parser.hh" // USES Parser
#include "spatialdata/geocoords/CSCart.hh" // HASA CoordSys
#include "spatialdata/geocoords/Converter.hh" // USES Converter
#include "spatialdata/muparser/muParser.h" // USES mu::parser

#include <vector> // USES std::vector
#include <stdexcept> // USES std::runtime_error

#include <sstream> // USES std::ostringsgream
#include <strings.h> // USES strcasecmp()
#include <assert.h> // USES assert()

// ----------------------------------------------------------------------
/// Default constructor
spatialdata::spatialdb::AnalyticDB::AnalyticDB(void) :
    _names(NULL),
    _scales(NULL),
    _expressions(NULL),
    _parsers(NULL),
    _cs(new spatialdata::geocoords::CSCart),
    _converter(new spatialdata::geocoords::Converter),
    _queryValues(NULL),
    _numValues(0),
    _querySize(0) {}


// ----------------------------------------------------------------------
/// Constructor with label
spatialdata::spatialdb::AnalyticDB::AnalyticDB(const char* label) :
    SpatialDB(label),
    _names(NULL),
    _scales(NULL),
    _expressions(NULL),
    _parsers(NULL),
    _cs(new spatialdata::geocoords::CSCart),
    _converter(new spatialdata::geocoords::Converter),
    _queryValues(NULL),
    _numValues(0),
    _querySize(0) {}


// ----------------------------------------------------------------------
/// Default destructor
spatialdata::spatialdb::AnalyticDB::~AnalyticDB(void) {
    clear();
    delete _cs;_cs = NULL;
    delete _converter;_converter = NULL;
} // destructor


// ----------------------------------------------------------------------
/// Clear database values.
void
spatialdata::spatialdb::AnalyticDB::clear(void) {
    delete[] _names;_names = NULL;
    delete[] _scales;_scales = NULL;
    delete[] _expressions;_expressions = NULL;
    delete[] _parsers;_parsers = NULL;
    delete[] _queryValues;_queryValues = NULL;
    _querySize = 0;
    _numValues = 0;
} // destructor


// ----------------------------------------------------------------------
// Set values in database.
void
spatialdata::spatialdb::AnalyticDB::setData(const char* const* names,
                                            const char* const* units,
                                            const char* const* expressions,
                                            const size_t numValues) {
    assert( (0 < numValues && names && units && expressions) ||
            (0 == numValues && !names && !units && !expressions) );

    clear();

    if (0 == numValues) {
        return;
    } // if

    spatialdata::units::Parser parser;

    _numValues = numValues;

    _names = new std::string[_numValues];
    for (size_t i = 0; i < _numValues; ++i) {
        _names[i] = names[i];
    } // for

    _scales = new double[_numValues];
    for (size_t i = 0; i < _numValues; ++i) {
        if (strcasecmp(units[i], "none") != 0) {
            _scales[i] = parser.parse(units[i]);
        } else {
            _scales[i] = 1.0;
        } // if/else
    } // for

    _expressions = new std::string[_numValues];
    _parsers = new mu::Parser[_numValues];
    for (size_t i = 0; i < _numValues; ++i) {
        _expressions[i] = expressions[i];
        _parsers[i].SetArgSep(',');
        _parsers[i].SetDecSep('.');
        // Only allow built-in variables mupSetVarFactory(_parsers, AddVariable, NULL);
        _parsers[i].DefineConst("pi", M_PI);
        _parsers[i].DefineVar("x", &_expVars[0]);
        _parsers[i].DefineVar("y", &_expVars[1]);
        _parsers[i].DefineVar("z", &_expVars[2]);
        _parsers[i].SetExpr(_expressions[i]);
    } // for

    // Default query values is all values.
    _querySize = _numValues;
    delete[] _queryValues;_queryValues = (_querySize > 0) ? new size_t[_querySize] : NULL;
    for (size_t i = 0; i < _querySize; ++i) {
        _queryValues[i] = i;
    } // for

} // setData


// ----------------------------------------------------------------------
// Set filename containing data.
void
spatialdata::spatialdb::AnalyticDB::setCoordSys(const geocoords::CoordSys& cs) {
    delete _cs;_cs = cs.clone();assert(_cs);
} // setCoordSys


// ----------------------------------------------------------------------
// Get names of values in spatial database.
void
spatialdata::spatialdb::AnalyticDB::getNamesDBValues(const char*** valueNames,
                                                     size_t* numValues) const {
    if (valueNames) {
        *valueNames = (_numValues > 0) ? new const char*[_numValues] : NULL;
        for (size_t i = 0; i < _numValues; ++i) {
            (*valueNames)[i] = _names[i].c_str();
        } // for
    }
    if (numValues) {
        *numValues = _numValues;
    } // if
} // getNamesDBValues


// ----------------------------------------------------------------------
// Set values to be returned by queries.
void
spatialdata::spatialdb::AnalyticDB::setQueryValues(const char* const* names,
                                                   const size_t numVals) {
    if (0 == numVals) {
        std::ostringstream msg;
        msg << "Number of values for query in spatial database " << getDescription()
            << "\n must be positive.\n";
        throw std::invalid_argument(msg.str());
    } // if
    assert(names && 0 < numVals);

    _querySize = numVals;
    delete[] _queryValues;_queryValues = new size_t[numVals];assert(_queryValues);
    for (size_t iVal = 0; iVal < numVals; ++iVal) {
        size_t iName = 0;
        while (iName < _numValues) {
            if (0 == strcasecmp(names[iVal], _names[iName].c_str())) {
                break;
            }
            ++iName;
        } // while
        if (iName >= _numValues) {
            std::ostringstream msg;
            msg << "Could not find value '" << names[iVal] << "' in spatial database '"
                << getDescription() << "'. Available values are:";
            for (size_t iName = 0; iName < _numValues; ++iName) {
                msg << "\n  " << _names[iName];
                msg << "\n";
                throw std::out_of_range(msg.str());
            } // for
        } // if
        _queryValues[iVal] = iName;
    } // for
} // queryVals


// ----------------------------------------------------------------------
// Query the database.
int
spatialdata::spatialdb::AnalyticDB::query(double* vals,
                                          const size_t numVals,
                                          const double* coords,
                                          const size_t numDims,
                                          const spatialdata::geocoords::CoordSys* csQuery) {
    if (0 == _querySize) {
        std::ostringstream msg;
        msg << "Values to be returned by spatial database " << getDescription() << "\n"
            << "have not been set. Please call setQueryValues() before query().\n";
        throw std::logic_error(msg.str());
    } // if
    else if (numVals != _querySize) {
        std::ostringstream msg;
        msg << "Number of values to be returned by spatial database "
            << getDescription() << "\n"
            << "(" << _querySize << ") does not match size of array provided ("
            << numVals << ").\n";
        throw std::invalid_argument(msg.str());
    } // if

    // Convert coordinates
    assert(numDims <= 3);
    for (size_t d = 0; d < numDims; ++d) {
        _expVars[d] = coords[d];
    }
    assert(csQuery);
    assert(_cs);
    assert(_converter);
    _converter->convert(_expVars, 1, numDims, _cs, csQuery);

    assert(_queryValues);
    assert(_scales);
    assert(_parsers);
    try {
        for (size_t iVal = 0; iVal < _querySize; ++iVal) {
            const size_t index = _queryValues[iVal];
            vals[iVal] = _scales[index] * _parsers[index].Eval();
        } // for
    } catch (const mu::Parser::exception_type& exception) {
        throw std::runtime_error(exception.GetMsg());
    }

    return 0;
} // query


// End of file

File: spatialdata/spatialdb/AnalyticDB.py
# =================================================================================================
# This code is part of SpatialData, developed through the Computational Infrastructure
# for Geodynamics (https://github.com/geodynamics/spatialdata).
#
# Copyright (c) 2010-2024, University of California, Davis and the SpatialData Development Team.
# All rights reserved.
#
# See https://mit-license.org/ and LICENSE.md and for license information. 
# =================================================================================================


from .SpatialDBObj import SpatialDBObj
from .spatialdb import AnalyticDB as ModuleAnalyticDB


class AnalyticDB(SpatialDBObj, ModuleAnalyticDB):
    """
    Spatial database composed of analytic functions.

    Implements `SpatialDB`.
    """
    DOC_CONFIG = {
        "cfg": """
            [db]
            description = Uniform material properties
            values = [density, vp, vs]
            units = [kg/m**3, km/s, m/s]
            expressions = [2.0*x-y, 2.0x*x/(y*y), -2.0*z]
            cs = spatialdata.geocoords.CSCart
            cs.space_dim = 3
            """,
    }

    import pythia.pyre.inventory

    values = pythia.pyre.inventory.list("values", default=[])
    values.meta['tip'] = "Names of values in spatial database."

    units = pythia.pyre.inventory.list("units", default=[])
    units.meta['tip'] = "Units for values in spatial database."

    expressions = pythia.pyre.inventory.list("expressions", default=[])
    expressions.meta['tip'] = "Analytical expressions for values in spatial database."

    from spatialdata.geocoords.CSCart import CSCart
    cs = pythia.pyre.inventory.facility("coordsys", factory=CSCart)
    cs.meta['tip'] = "Coordinate system."

    def __init__(self, name="AnalyticDB"):
        """
        Constructor.
        """
        SpatialDBObj.__init__(self, name)
        from pythia.pyre.units import parser
        self.parser = parser()

    def _configure(self):
        """
        Set members based on inventory.
        """
        SpatialDBObj._configure(self)
        self._validateParameters(self.inventory)
        self.cs._configure()
        ModuleAnalyticDB.setCoordSys(self, self.cs)
        values = []
        for v in self.values:
            values.append(v.strip())
        ModuleAnalyticDB.setData(self, values, self.units, self.expressions)

    def _createModuleObj(self):
        """
        Create Python module object.
        """
        ModuleAnalyticDB.__init__(self)

    def _validateParameters(self, params):
        """
        Validate parameters.
        """
        if len(params.values) == 0:
            raise ValueError("Values in AnalyticDB '%s' not specified.", self.description)
        if len(params.units) == 0:
            raise ValueError("Units for AnalyticDB '%s' not specified." % self.description)
        if len(params.expressions) == 0:
            raise ValueError("Analytical expressions for AnalyticDB '%s' not specified." % self.description)
        if len(params.values) != len(params.units or len(params.values) != len(params.expressions)):
            raise ValueError("Incompatible settings for uniform spatial database '%s'.\n"
                             "'values', 'units', and 'expressions' must be lists of the same size.\n"
                             "Size of 'values'=%d, 'units'=%d, and 'expressions'=%d."
                             % (self.description, len(params.values), len(params.units), len(params.expressions)))
        try:
            for x in params.units:
                if len(str(x).split("*")) > 1:
                    xdim = self.parser.parse(str(x))
        except:
            raise ValueError(
                "'units' for AnalyticDB '%s' must contain dimensioned or nondimensional values." % self.description)


def spatial_database():
    """
    Factory associated with AnalyticDB.
    """
    return AnalyticDB()


# End of file
